packages/libs/error-stack/src/fmt/mod.rs

```diff
@@ -0,0 +1,1075 @@
+//! Implementation of formatting, to enable colors and the use of box-drawing characters use the
+//! `pretty-print` feature.
+//!
+//! # Hooks
+//!
+//! The format implementation (especially the [`Debug`] implementation),
+//! can be easily extended using hooks.
+//! Hooks are functions of the signature `Fn(&T, &mut HookContext<T>) -> Emit`, they provide an easy
+//! and ergonomic way to partially modify the format and enable the output of types that are not
+//! necessarily added via `.attach_printable()` or are unable to implement [`Display`].
+//!
+//! Hooks can be attached through the central hooking mechanism which `error-stack`
+//! provides via [`Report::install_debug_hook`].
+//!
+//! You can also provide a fallback function, which is called whenever a hook hasn't been added for
+//! a specific type of attachment.
+//! The fallback function needs to have a signature of
+//! `Fn(&Frame, HookContext<'a, T>) -> Call<'a, Frame>`
+//! and can be set via [`Report::install_debug_hook_fallback`].
+//!
+//! > **Caution:** Overwriting the fallback **will** remove the builtin formatting for types like
+//! > [`Backtrace`] and [`SpanTrace`], you can mitigate this by calling
+//! > [`error_stack::fmt::builtin`] in your fallback code.
+//!
+//! Hook functions need to be [`Fn`] and **not** [`FnMut`], which means they are unable to directly
+//! mutate state outside of the closure.
+//! You can still achieve mutable state outside of the scope of your closure through interior
+//! mutability, e.g. by using the [`std::sync`] module like [`Mutex`], [`RwLock`], and [`atomic`]s.
+//!
+//! The type a hook will be called for is determined by the type of the first argument.
+//! This type can either be specified at the closure level or when calling
+//! [`Report::install_debug_hook`].
+//! This type needs to be `'static`, [`Send`], and [`Sync`].
```

Comment:
I don't think this needs to be stated, as you cannot attach any object, which is not `'static + Send + Sync`. Being explicit about this however does not hurt